client/components/webpack-build-monitor/index.tsx

```diff
@@ -0,0 +1,222 @@
+/**
+ * @see https://webpack.js.org/api/module-variables/#__webpack_hash__-webpack-specific
+ */
+declare const __webpack_hash__: string;
+
+/**
+ * External dependencies
+ */
+import React, { FunctionComponent, useEffect, useState } from 'react';
+import classNames from 'classnames';
+import debugFactory from 'debug';
+import io from 'socket.io-client';
+
+/**
+ * Internal dependencies
+ */
+import Spinner from 'components/spinner';
+import { IntervalHandle, TimestampMS } from 'client/types';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+const CONNECTION_TIMEOUT = 20 * 1000;
+const debug = debugFactory( 'calypso:webpack-build-monitor' );
+
+const WebpackBuildMonitor: FunctionComponent = () => {
+	const [ lastHash, setLastHash ] = useState( __webpack_hash__ );
+
+	// Webpack
+	const [ isWebpackConnected, setIsWebpackConnected ] = useState( false );
+	const [ hasWebpackErrors, setHasWebpackErrors ] = useState( false );
+	const [ hasWebpackWarnings, setHasWebpackWarnings ] = useState( false );
+	const [ isWebpackBuilding, setWebpackBuilding ] = useState( false );
+
+	// CSS
+	const [ isCssConnected, setIsCssConnected ] = useState( false );
+	const [ hasCssErrors, setHasCssErrors ] = useState( false );
+	const [ isCssBuilding, setCssBuilding ] = useState( false );
+
+	useEffect( () => {
+		const namespace = location.protocol + '//' + location.host + '/css-hot-reload';
+
+		let socket: SocketIOClient.Socket | undefined;
+
+		const connect = () => {
+			debug( 'Hot CSS connecting' );
+			socket = io.connect( namespace, {
+				reconnectionDelay: CONNECTION_TIMEOUT,
+				transports: [ 'websocket' ],
+			} );
+			socket.on( 'connect', () => {
+				debug( 'Hot CSS connected' );
+				setIsCssConnected( true );
+			} );
+
+			socket.on( 'disconnect', () => {
+				debug( 'Hot CSS disconnected. Reconnecting…' );
+				setIsCssConnected( false );
+			} );
+
+			socket.on(
+				'css-hot-reload',
+				( { status }: { status: 'building' | 'build-failed' | 'reload' | 'up-to-date' } ) => {
+					debug( 'Hot CSS status update: %o', status );
+					switch ( status ) {
+						case 'building':
+							setCssBuilding( true );
+							break;
+
+						case 'build-failed':
+							setHasCssErrors( true );
+							setCssBuilding( false );
+							break;
+
+						case 'reload':
+						case 'up-to-date':
+							setHasCssErrors( false );
+							setCssBuilding( false );
+							break;
+					}
+				}
+			);
+		};
+
+		connect();
+
+		return () => {
+			if ( socket && ! socket.disconnected ) {
+				socket.close();
+			}
+		};
+	}, [] );
+
+	useEffect( () => {
+		if ( typeof EventSource === 'undefined' ) {
+			if ( process.env.NODE_ENV !== 'production' ) {
+				// eslint-disable-next-line no-console
+				console.warn( 'Webpack build monitor disabled. No `EventSource`.' );
+			}
+		} else {
+			let lastActivity: TimestampMS;
+			let connectionTimer: IntervalHandle;
+			let source: EventSource;
+			const connect = () => {
```

Comment:
I guess there's no pragmatic reason to have these next few lines in a function right? Just stylistic (to organise the code)?

Reply:
If the connection times out, we reconnect:

https://github.com/Automattic/wp-calypso/pull/35839/files#diff-a6c37081699f0ac5686902eceffdcda4R120

Reply:
Yup, gotcha!